### DISCUSSION TITLE
Is the PlanarEmbedding() functionality correct? How to use it?

### DISCUSSION THREAD

**Discussion by TheRolf** [networkx/networkx#5079] [Q&A]
There is only a very small example in the documentation (or the whole internet) about defining a planar embedding. Based on that, it is not clear how to do it for any (planar) graph, I didn't manage to make it work for a triangle graph, after an hour of trial-and-error. Why is the following not correct?

```
P = nx.PlanarEmbedding()

P.add_half_edge_cw(0, 1, None)
P.add_half_edge_cw(0, 2, 1)
P.add_half_edge_cw(1, 2, None)

P.add_half_edge_cw(1, 0, None)
P.add_half_edge_cw(2, 0, None)
P.add_half_edge_cw(2, 1, 0)
```

Furthermore, I am not sure this functionality always gives the correct embedding. If I try for example

```
P = nx.PlanarEmbedding()

P.add_half_edge_cw(0, 1, None)
P.add_half_edge_cw(0, 2, 1)
P.add_half_edge_cw(0, 3, 2)
P.add_half_edge_cw(0, 4, 3)

P.add_half_edge_cw(1, 0, None)
P.add_half_edge_cw(2, 0, None)
P.add_half_edge_cw(3, 0, None)
P.add_half_edge_cw(4, 0, None)
```

it gives a star graph with the leaves in the order 1, 4, 2, 3; instead of 1, 2, 3, 4. Isn't the definition of the half-edges above saying _"2 is a neighbour of 0, right of 1"_ and _"3 is a neighbour of 0, right of 2"_?

**Comment by dschult** [comment]
If you add the line `P.check_structure()` after your first example it reports that the graph structure has a "bad embedding. The edge orientations are not correct."   I think the trouble is that you should only use `None` as the reference_neighbor when the starting node has no neighbors. 

The second example produces an object P with valid structure. But when I do it the edges are in the order 1,2,3,4 as expected. How are you reporting the order of the leaves?
```
>>> P.edges(data=True)
OutEdgeDataView([(0, 1, {'cw': 2, 'ccw': 4}), (0, 2, {'cw': 3, 'ccw': 1}), (0, 3, {'cw': 4, 'ccw': 2}), (0, 4, {'cw': 1, 'ccw': 3}), (1, 0, {'cw': 0, 'ccw': 0}), (2, 0, {'cw': 0, 'ccw': 0}), (3, 0, {'cw': 0, 'ccw': 0}), (4, 0, {'cw': 0, 'ccw': 0})])
```

**Comment by dschult** [comment]
The best way to get the planar embedding for a planar graph is to use the function from that same module:
```python
# suppose G is a planar graph
(foo, P) = nx.check_planarity(G)
if foo:
    assert P.check_structure()
else:
    print("G is not planar")
```

**Comment by TheRolf** [comment]
Thank you for your replies!

I did try the `check_structure()` for verifying, but the error messages were not really helpful as in how to fix the embedding. I didn't notice the `check_planarity()` actually gives you a planar embedding if the graph is planar, and there is no need for manually creating it.

What I ended up doing is to add the faces of the graph manually, edge by edge to a planar embedding P, setting 'cw' reference neighbour if the target of the edge is new, and 'ccw' if it already exists in P. I guess this is still useful to have if I want a *specific* planar embedding.

As for the second example, what I meant is that the `draw()` method seems to choose a random order for the neighbours. But I guess the "clockwise ordering" of the neighbours is not necessarily the same to the visual order.

**Reply by dschult** [reply]
Correct -- the drawing functions do not know how to use the information from the planar embedding.
Looks like we should probably update the documentation to make it clear that the `check_planarity` function is the primary interface for the planar embedding.

**Reply by TheRolf** [reply]
I agree, I think it would be helpful for future users!
